spec/tapioca/runtime/generic_type_registry_spec.rb

```diff
@@ -0,0 +1,103 @@
+# typed: true
+# frozen_string_literal: true
+
+require "spec_helper"
+
+module Tapioca
+  module Runtime
+    class GenericTypeRegistrySpec < Minitest::Spec
+      describe Tapioca::Runtime::GenericTypeRegistry do
+        describe ".generic_type_instance?" do
+          it "returns false for instances of non-generic classes" do
+            refute(GenericTypeRegistry.generic_type_instance?(Object.new))
+          end
+
+          it "returns true for instances of generic classes" do
+            assert(GenericTypeRegistry.generic_type_instance?(SampleGenericClass[Object].new))
+          end
+        end
+
+        describe ".lookup_type_variables" do
+          it "returns nil for non-generic types" do
+            assert_nil(GenericTypeRegistry.lookup_type_variables(Object))
+          end
+
+          it "returns the type variables for generic types" do
+            assert_equal([SampleGenericClass::Element], GenericTypeRegistry.lookup_type_variables(SampleGenericClass))
+          end
+        end
+
+        describe ".register_type_variable" do
+          it " registers a type variable that can be looked up later" do
+            not_actually_generic = Class.new
+
+            fake_type_member1 = Tapioca::TypeVariableModule.new(
+              not_actually_generic,
+              Tapioca::TypeVariableModule::Type::Member,
+              :invariant,
+              nil,
+              nil,
+              nil,
+              nil,
+            )
+
+            fake_type_member2 = Tapioca::TypeVariableModule.new(
+              not_actually_generic,
+              Tapioca::TypeVariableModule::Type::Member,
+              :invariant,
+              nil,
+              nil,
+              nil,
+              nil,
+            )
+
+            GenericTypeRegistry.register_type_variable(not_actually_generic, fake_type_member1)
+            GenericTypeRegistry.register_type_variable(not_actually_generic, fake_type_member2)
+
+            type_variables = T.must(GenericTypeRegistry.lookup_type_variables(not_actually_generic))
+
+            assert_equal([fake_type_member1, fake_type_member2], type_variables)
+            # Let's double-check that they're not just equal, but identical:
+            assert_same(fake_type_member1, type_variables[0])
+            assert_same(fake_type_member2, type_variables[1])
+          end
+        end
+
+        describe "the patch for .inherited on generic classes" do
+          # This is more of an internal detail and cross-cutting concern of all the public APIs,
+          # but it's easier to test here on its own.
+
+          it "rescues exceptions that would have prevents subclassing" do
+            assert_raises(RuntimeError) do
+              # Precondition. If not caught by the patch, this error would have prevented our subclassing.
+              RaisesInInheritedCallback.inherited(Class.new)
```

Comment:
Same thing here, can we nest the test class inside the test?

Reply:
Oh yeah, this one only exists for this test case. Done!

Reply:
Ah wait, Sorbet doesn't like this, and I'm not sure how to convince it haha

Here's my attempt: https://github.com/Shopify/tapioca/commit/838d4ffefa07e9f4582e4732dd6c2dbf01575276

```
spec/tapioca/runtime/generic_type_registry_spec.rb:74: Method type_member does not exist on T::Class[Object] https://srb.help/7003
    74 |              Element = type_member
                                ^^^^^^^^^^^
  Got T::Class[Object] originating from:
    spec/tapioca/runtime/generic_type_registry_spec.rb:71:
    71 |            cls = Class.new do
    72 |              extend T::Generic
    73 |
    74 |              Element = type_member
    75 |
       |...
    78 |                  super
    79 |                  raise "Boom"
    80 |                end
    81 |              end
    82 |            end

spec/tapioca/runtime/generic_type_registry_spec.rb:89: Method [] does not exist on T::Class[Object] https://srb.help/7003
    89 |            result = cls[Object] # Should be a distinct subclass
                             ^^^^^^^^^^^
  Got T::Class[Object] originating from:
    spec/tapioca/runtime/generic_type_registry_spec.rb:71:
    71 |            cls = Class.new do
    72 |              extend T::Generic
    73 |
    74 |              Element = type_member
    75 |
       |...
    78 |                  super
    79 |                  raise "Boom"
    80 |                end
    81 |              end
    82 |            end
Errors: 2
```